dhcp-relay/dhcp_kern_xdp.c

```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
+
+		}
+
+		if (vlans.id[1] != 0) {
+			bpf_printk("Inner VLAN %i found, will change to %i", vlans.id[1], new_vlans->id[1]);
+
+			struct vlan_hdr *inner_vlh = data + ETH_HLEN + sizeof (struct vlan_hdr);
+			inner_vlh->h_vlan_TCI = bpf_htons((bpf_ntohs(inner_vlh->h_vlan_TCI) & 0xf000) | new_vlans->id[1]);
+
+		} else {
+			bpf_printk("Inner VLAN not found, will insert %i", new_vlans->id[1]);
+
+			/* Adjust header by -4 bytes to make space for VLAN header */
+			if (bpf_xdp_adjust_head(ctx, -(int) sizeof (struct vlan_hdr))) {
+				bpf_printk("Cannot head-adjust packet by %i bytes, aborting", -(int) sizeof (struct vlan_hdr));
+				rc = XDP_ABORTED;
+				goto out;
+			}
+
+			bpf_printk("Head-adjusted packet by %i bytes", -(int) sizeof (struct vlan_hdr));
+
+			head_adjusted = 1;
+
+			data_end = (void *) (long) ctx->data_end;
+			data = (void *) (long) ctx->data;
+
+			/* Verifier check */
+			if (data + ETH_ALEN + ETH_ALEN + sizeof (struct vlan_hdr) + sizeof (struct vlan_hdr) > data_end) {
+				rc = XDP_ABORTED;
+				goto out;
+			}
+
+			/* Move MAC address headers + outer VLAN tag to beginning of packet */
+			memmove(data, data + sizeof (struct vlan_hdr), ETH_ALEN + ETH_ALEN + sizeof (struct vlan_hdr));
+
+			bpf_printk("Moved %i bytes from offset %i to offset %i", ETH_ALEN + ETH_ALEN + sizeof (struct vlan_hdr), sizeof (struct vlan_hdr), 0);
+
+			/* Make new inner VLAN header (copy from outer VLAN header) */
+			memcpy(data + ETH_ALEN + ETH_ALEN + sizeof (struct vlan_hdr), data + ETH_ALEN + ETH_ALEN, sizeof (struct vlan_hdr));
+
+			bpf_printk("Copied %i bytes from offset %i to offset %i", sizeof (struct vlan_hdr), ETH_ALEN + ETH_ALEN, ETH_ALEN + ETH_ALEN + sizeof (struct vlan_hdr));
+
+			bpf_printk("Will modify VLAN header at offset %i", ETH_ALEN + ETH_ALEN + sizeof (struct vlan_hdr));
+
+			struct vlan_hdr *vlh = data + ETH_ALEN + ETH_ALEN + sizeof (struct vlan_hdr) + 2;
+			vlh->h_vlan_TCI = bpf_htons((bpf_ntohs(vlh->h_vlan_TCI) & 0xf000) | new_vlans->id[1]);
+
+			offset += sizeof (struct vlan_hdr);
+			vlan_length += sizeof (struct vlan_hdr);
+
+			/* Parse DHCP packet */
+			if (data + vlan_length + dhcp_offset + sizeof (dhcp) > data_end) {
+				goto out;
+			}
+			dhcp = data + vlan_length + dhcp_offset;
+
+			bpf_printk("Inserted VLAN header");
+
+		}
+
+		rc = XDP_TX;
+
+	}
 
 	/* Check if we exceed boundaries to make verifier happy */
 	if (data + offset > data_end)
 		goto out;
 
 	option_offset = offset;
 
+	__u8 n = 0;
+
 	__u8 *pos = (__u8 *) (data + option_offset);
 
 	/* Loop through all DHCP options */
-#pragma unroll DHCP_MAX_OPTIONS
+	//#pragma unroll DHCP_MAX_OPTIONS
 	for (i = 0; i < DHCP_MAX_OPTIONS; i++) {
 
 		/* Verifier check */
 		if (pos + 1 > data_end)
 			break;
 
+		/* Read option code */
 		option_code = *pos;
 
 		bpf_printk("Got option code %i at offset %i, hex %x", option_code, option_offset, option_offset);
 
+		if (option_code == 82 && dhcp->op == DHCP_REPLY) {
```

Comment:
This code is wrong for several reasons:

- It doesn't handle option 82 for REQUESTS (it should drop the packet)
- It just blindly turns option 82 into an option 255, regardless of where it is in the packet (what if it's not actually the last option)
- It assumes that the option 82 is always the same size as the one we put in
- It leaves junk in the packet

The right thing to do is to shrink the packet; and either move everything up, including any non-255 options after this, or at least checking to see if there is anything other than an end option in the rest of the packet (and dropping it as invalid if there is).